### PR TITLE
test: increase test suite coverage to 100% (#156)

### PR DESCRIPTION
**Pull request by x48-crypto**
This PR brings test suite coverage to 100%. It should be noted that BaseStrategy is under-reporting due to a Brownie error related to modifiers (https://github.com/eth-brownie/brownie/issues/809), however I've ensured that even the methods that report as not fully covered are indeed covered.

I've rebased the PR against @snorkr4's prior work as well (https://github.com/iearn-finance/yearn-vaults/pull/200).

IMAGE 2021-03-01 03:59:07
IMAGE 2021-03-01 03:59:09

### CODE REVIEW

File: brownie-config.yaml [hunk 1]
```diff
@@ -21,3 +21,5 @@ reports:
     - contracts/test/TestGuestList.sol
   exclude_contracts:
     - SafeMath
+    - SafeERC20
+    - Address
```
Comment (x48-crypto) [inline brownie-config.yaml:25]: Removed SafeERC20 and Address from coverage report as they are not a Yearn contracts (they're from OpenZeppelin)

File: contracts/Vault.vy [hunk 2]
```diff
@@ -444,7 +444,7 @@ def setLockedProfitDegration(degration: uint256):
     @param degration The rate of degration in percent per second scaled to 1e18.
     """
     assert msg.sender == self.governance
-    assert degration >= 0 and degration <= DEGREDATION_COEFFICIENT
+    assert degration <= DEGREDATION_COEFFICIENT
```
Comment (x48-crypto) [inline contracts/Vault.vy:447]: Due to the way the this is structured it's not possible to obtain full coverage for this line. Since "degration" is a uint256 it can never be less than zero (you will get an outside of range error if you try). The only way to obtain full coverage here is to change the assert statement in Vault.vy.

File: contracts/Vault.vy [hunk 3]
```diff
@@ -1331,13 +1331,16 @@ def addStrategyToQueue(strategy: address):
     assert msg.sender in [self.management, self.governance]
     # Must be a current Strategy
     assert self.strategies[strategy].activation > 0
-    # Check if queue is full
-    assert self.withdrawalQueue[MAXIMUM_STRATEGIES - 1] == ZERO_ADDRESS
     # Can't already be in the queue
+    last_idx: uint256 = 0
```
Comment (x48-crypto) [inline contracts/Vault.vy:1335]: Updated strategy queue code per @fubuloubu . New code is slightly more gas efficient and is fully testable

File: contracts/test/TestStrategy.sol [hunk 4]
```diff
@@ -25,13 +26,18 @@ contract TestStrategy is BaseStrategyInitializable {
 
     // NOTE: This is a test-only function to simulate losses
     function _takeFunds(uint256 amount) public {
-        want.transfer(msg.sender, amount);
+        want.safeTransfer(msg.sender, amount);
```
Comment (x48-crypto) [inline contracts/test/TestStrategy.sol:29]: This change is already merged

File: contracts/test/Token.sol [hunk 5]
```diff
@@ -23,20 +24,73 @@ contract Token is ERC20 {
         require(!_blocked[to], "Token transfer refused. Receiver is on blacklist");
         super._beforeTokenTransfer(from, to, amount);
     }
+}
+
+contract TokenNoReturn {
```
Comment (x48-crypto) [inline contracts/test/Token.sol:29]: This is required to get full coverage on erc20_safe_transfer and erc20_safe_transferFrom

File: tests/functional/conftest.py [hunk 6]
```diff
@@ -1,5 +1,7 @@
 import pytest
 
+from brownie import Token, TokenNoReturn
```
Comment (x48-crypto) [inline tests/functional/conftest.py:3]: These changes are already merged to master

File: tests/functional/strategy/test_misc.py [hunk 7]
```diff
@@ -81,6 +81,39 @@ def test_harvest_tend_trigger(chain, gov, vault, token, TestStrategy):
     assert strategy.estimatedTotalAssets() == 0
 
 
+def test_sweep_authority(
```
Comment (x48-crypto) [inline tests/functional/strategy/test_misc.py:84]: This test is for the "authority" modifier, although the test is not correctly picked up by brownie due to https://github.com/eth-brownie/brownie/issues/809

File: tests/functional/vault/test_config.py [hunk 8]
```diff
@@ -52,14 +53,25 @@ def test_vault_name_symbol_override(guardian, gov, rewards, token, Vault):
...
     "getter,setter,val,guard_allowed",
     [
         ("name", "setName", "NewName yVault", False),
         ("symbol", "setSymbol", "yvNEW", False),
         ("emergencyShutdown", "setEmergencyShutdown", True, True),
+        ("emergencyShutdown", "setEmergencyShutdown", False, False),
         ("guardian", "setGuardian", None, True),
+        ("guestList", "setGuestList", None, False),
         ("rewards", "setRewards", None, False),
+        ("lockedProfitDegration", "setLockedProfitDegration", 1000, False),
```
Comment (x48-crypto) [inline tests/functional/vault/test_config.py:74]: I believe this was added by @snorkr4

File: contracts/Vault.vy (hunk 2 above)
Comment (fubuloubu) [reply contracts/Vault.vy:447]: Add a comment above this

File: contracts/Vault.vy (hunk 3 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1335]: Should result in 1 less `SLOAD` (note currently iteration over `self.withdrawalQueue` just loads the whole thing anyways)

File: contracts/test/TestStrategy.sol [hunk 9]
```diff
@@ -85,8 +91,7 @@ contract TestStrategy is BaseStrategyInitializable {
         } else {
             // NOTE: Just in case something was stolen from this contract
             if (totalDebt > totalAssets) {
-                _loss = totalDebt.sub(totalAssets);
-                if (_loss > _amountNeeded) _loss = _amountNeeded;
+                _loss = _amountNeeded;
```
Comment (fubuloubu) [inline contracts/test/TestStrategy.sol:94]: I don't think this is correct

File: tests/functional/vault/test_config.py [hunk 10]
```diff
@@ -177,3 +189,11 @@ def test_vault_setGovernance(gov, vault, rando):
...
     with brownie.reverts():
         vault.acceptGovernance({"from": gov})
+
+
+def test_vault_setLockedProfitDegration_range(gov, vault):
+    # value must be between 0 and DEGREDATION_COEFFICIENT (inclusive)
+    vault.setLockedProfitDegration(0, {"from": gov})
+    vault.setLockedProfitDegration(DEGREDATION_COEFFICIENT, {"from": gov})
+    with brownie.reverts():
+        vault.setLockedProfitDegration(DEGREDATION_COEFFICIENT + 1000, {"from": gov})
```
Comment (fubuloubu) [inline tests/functional/vault/test_config.py:199]: Show off-by-one behavior correct
```suggestion
        vault.setLockedProfitDegration(DEGREDATION_COEFFICIENT + 1, {"from": gov})
```

File: tests/functional/vault/test_misc.py [hunk 11]
```diff
@@ -2,12 +2,97 @@
...
 def other_token(gov, Token):
     yield gov.deploy(Token, 18)
 
 
+@pytest.fixture
+def token_false_return(gov, TokenFalseReturn):
+    yield gov.deploy(TokenFalseReturn, 18)
+
+
+def vault(gov, management, token, Vault):
```
Comment (fubuloubu) [inline tests/functional/vault/test_misc.py:19]: Missing `fixture`?

File: tests/functional/vault/test_strategies.py [hunk 12]
```diff
@@ -30,29 +28,59 @@ def other_strategy(gov, vault, TestStrategy):
...
     # Only governance can add a strategy
     with brownie.reverts():
         vault.addStrategy(strategy, 100, 10, 20, 1000, {"from": rando})
 
+    # Can't add a strategy during emergency shutdown
+    vault.setEmergencyShutdown(True, {"from": gov})
+    with brownie.reverts():
+        vault.addStrategy(strategy, 100, 10, 20, 1000, {"from": gov})
+    chain.undo()
+    chain.undo()
```
Comment (fubuloubu) [inline tests/functional/vault/test_strategies.py:82]: Why undo twice? this puts it before the test?

Comment (snorkr4) [reply tests/functional/vault/test_strategies.py:82]: I can't find any documentation on this, but as far as I can tell from experimentation - at line 81, the `addStrategy` call from line 80 is still in the transaction history. That's why I added the second undo.

File: tests/functional/vault/test_strategies.py [hunk 13]
```diff
@@ -314,8 +354,42 @@ def test_ordering(gov, vault, TestStrategy, rando):
...
+
+    # Not just anyone can add a strategy to the queue
+    with brownie.reverts():
+      vault.addStrategyToQueue(strategy, {"from": rando})
+
+    # # Governance can add a strategy to the queue
+    vault.addStrategyToQueue(strategy, {"from": gov})
+    vault.removeStrategyFromQueue(strategy, {"from": gov})
+      
+    # Managemenet can add a strategy to the queue
```
Comment (snorkr4) [inline tests/functional/vault/test_strategies.py:375]: nit: typo
```suggestion
    # Management can add a strategy to the queue
```

File: tests/functional/vault/test_strategies.py [hunk 14]
```diff
@@ -314,8 +354,42 @@ def test_ordering(gov, vault, TestStrategy, rando):
     with brownie.reverts():
         vault.addStrategyToQueue(removed_strategy, {"from": gov})
 
+def test_add_strategy_to_queue(gov, management, vault, TestStrategy, strategy, other_strategy, rando):
```
Comment (snorkr4) [inline tests/functional/vault/test_strategies.py:357]: nit: to be stylistically consistent with `test_addStrategy`, I guess this could be called `test_addStrategyToQueue`?

File: tests/functional/vault/test_strategies.py (hunk 12 above)
Comment (fubuloubu) [reply tests/functional/vault/test_strategies.py:82]: I don't think that is the case, but if you were having issues with it we should at least add a comment here as to what is being rolled back with each call.

**Review by snorkr4 — COMMENTED** [review]
Thanks for tagging me and including my changes. Was helpful to me to read through this.

File: contracts/test/TestStrategy.sol (hunk 9 above)
Comment (x48-crypto) [reply contracts/test/TestStrategy.sol:94]: In my testing I've not been able to get `_loss > _amountNeeded` to return false. I believe this line is over-defined/always true, so I set `_loss = _amountNeeded` as this is what happens when `_loss > _amountNeeded` is true.

File: tests/functional/vault/test_strategies.py (hunk 12 above)
Comment (snorkr4) [inline tests/functional/vault/test_strategies.py:82]:
```suggestion
    # Even if the previous call to vault.addStrategy reverts, it does not 
    # rollback, so we must undo twice to undo the emergency shutdown.
    chain.undo() # vault.addStrategy
    chain.undo() # vault.setEmergencyShutdown
```

Comment (x48-crypto) [reply tests/functional/vault/test_strategies.py:82]: @snorkr4 what would you suggest for a comment here?

Comment (x48-crypto) [reply tests/functional/vault/test_strategies.py]: @snorkr4 nvm, saw your suggestion and merged. thanks!

File: tests/functional/vault/test_misc.py (hunk 11 above)
Comment (x48-crypto) [reply tests/functional/vault/test_misc.py:19]: @snorkr4 do you know why this is by any chance? I believe this came from your PR originally. I noticed if I add the fixture the test fails

File: contracts/test/TestStrategy.sol (hunk 9 above)
Comment (x48-crypto) [reply contracts/test/TestStrategy.sol:94]: I've reverted this commit as I was able to add a new unit test to obtain full coverage without modifying this code

image
image

**Comment by fubuloubu** [comment]
Fix commitlint and run `black` on changes

**Review by fubuloubu — APPROVED** [review]

File: tests/functional/vault/test_misc.py (hunk 11 above)
Comment (x48-crypto) [reply tests/functional/vault/test_misc.py:20]: I've made `vault` a fixture and fixed the broken unit tests

**Comment by x48-crypto** [comment]
All issues have been resolved and the PR has been rebased against master.

File: tests/functional/vault/test_misc.py [hunk 15]
```diff
@@ -49,18 +50,19 @@ def test_credit_available_minDebtPerHarvest_larger_than_available(
...
-    vault.setDepositLimit(2 ** 256 - 1, {"from": gov})
+    vault.setDepositLimit(MAX_UINT256, {"from": gov})
     strategy = gov.deploy(TestStrategy, vault)
+    minDebtPerHarvest = MAX_UINT256 - 1
     vault.addStrategy(
         strategy,
         10000,  # 100% of Vault AUM
-        2 ** 256 - 2,  # minDebtPerHarvest
-        2 ** 256 - 1,  # maxDebtPerHarvest
+        minDebtPerHarvest,  # minDebtPerHarvest
```
Comment (fubuloubu) [inline tests/functional/vault/test_misc.py:59]: why is this set so high? (can it be 0)?

Comment (x48-crypto) [reply tests/functional/vault/test_misc.py]: In order to test the line `available < strategy_minDebtPerHarvest` `minDebtToHarvest` must be greater than `available` (it cannot be zero).

Previously I had set `minDebtToHarvest` to `MAX_UINT256 - 1` to ensure this line gets tested correctly, but I've updated the PR and now I start with `minDebtToHarvest` set to zero and update it to `available + 1`.

It should be noted that this test is specifically to test that `creditAvailable` returns zero after a liquidation when `minDebtToHarvest` is greater than `available`. We do have another unit test that covers liquidation where `minDebtPerHarvest` is zero (or less than `available`).

This is the primary assert we are testing for:
```
    minDebtPerHarvestExceedsAvailable = strategy_minDebtPerHarvest > available
    assert minDebtPerHarvestExceedsAvailable == True

    creditAvalable = vault.creditAvailable(strategy)
    assert creditAvalable == 0
```

This is the code we are adding coverage for:
image

Comment (fubuloubu) [reply tests/functional/vault/test_misc.py]: Sounds good!
